File: backend/src/main/java/com/utsman/backend/controller/DriverController.java
package com.utsman.backend.controller;


import com.utsman.backend.model.Driver;
import com.utsman.backend.model.Message;
import com.utsman.backend.repository.DriverRepository;

import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

@RestController
@RequestMapping("/api/driver")
public class DriverController {

    @Autowired
    private DriverRepository driverRepository;

    @RequestMapping(value = "/", method = RequestMethod.GET)
    public List<Driver> getAllDriver() {
        return driverRepository.findAll();
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public Driver getDriver(@PathVariable("id") ObjectId id) {
        return driverRepository.findBy_id(id);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.PUT)
    public Driver editDriver(@PathVariable("id") ObjectId id, @Valid @RequestBody Driver driver) {
        driver.set_id(id);
        driverRepository.save(driver);
        return driver;
    }

    @RequestMapping(value = "/", method = RequestMethod.POST)
    public Driver addDriver(@Valid @RequestBody Driver driver) {
        driver.set_id(ObjectId.get());
        driverRepository.save(driver);
        return driver;
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.DELETE)
    public Message deleteDriver(@PathVariable("id") ObjectId id) {
        Driver driver = driverRepository.findBy_id(id);

        try {
            driverRepository.delete(driver);
            return new Message("Delete Success");
        } catch (IllegalArgumentException e) {
            return new Message("Delete failed, maybe driver cannot find or " + e.getLocalizedMessage());
        }
    }
}

File: backend-side/src/main/kotlin/com/kemana/backend/controller/DriverController.kt
package com.kemana.backend.controller

import com.kemana.backend.model.Position
import com.kemana.backend.model.Responses
import com.kemana.backend.model.Driver
import com.kemana.backend.repository.DriverRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import java.util.*
import javax.validation.Valid

@RestController
@RequestMapping("/api/v1/driver")
class DriverController {
    @Autowired
    lateinit var driverRepository: DriverRepository

    @RequestMapping(value = ["/active"], method = [RequestMethod.GET])
    fun getActiveDriver(): Responses {
        val drivers = driverRepository.findAll()
        return Responses("OK", drivers)
    }

    @RequestMapping(value = ["/"], method = [RequestMethod.POST])
    fun saveDriver(@Valid @RequestBody driver: Driver): Responses {
        driver.id = UUID.nameUUIDFromBytes(driver.email.toByteArray()).toString()
        driverRepository.save(driver)
        return Responses("OK", listOf(driver))
    }

    @RequestMapping(value = ["/{id}"], method = [RequestMethod.GET])
    fun getDriver(@PathVariable("id") id: String): Responses {
        val driver = driverRepository.findDriverById(id)
        return Responses("OK", listOf(driver))
    }

    @RequestMapping(value = [""], method = [RequestMethod.GET])
    fun getDriverByEmail(@RequestParam("email") email: String): Responses {
        val driver = driverRepository.findDriverByEmail(email)
        return Responses("OK", listOf(driver))
    }

    @RequestMapping(value = [""], method = [RequestMethod.PUT])
    fun editPositionByEmail(@RequestParam("email") email: String, @Valid @RequestBody position: Position): Responses {
        val driver = driverRepository.findDriverByEmail(email)
        driver?.position = position
        driverRepository.save(driver!!)
        return Responses("OK", listOf(driver))
    }

    @RequestMapping(value = ["/{id}"], method = [RequestMethod.PUT])
    fun editPosition(@PathVariable("id") id: String, @Valid @RequestBody position: Position): Responses {
        val driver = driverRepository.findDriverById(id)
        driver?.position = position
        driverRepository.save(driver!!)
        return Responses("OK", listOf(driver))
    }

    @RequestMapping(value = [""], method = [RequestMethod.DELETE])
    fun deleteByEmail(@RequestParam("email") email: String): Responses {
        val driver = driverRepository.findDriverByEmail(email)

        return if (driver != null) {
            driver.let { driverRepository.delete(it) }
            Responses("OK", listOf(null))
        } else {
            Responses("FAILED", listOf(null))
        }
    }

    @RequestMapping(value = ["/{id}"], method = [RequestMethod.DELETE])
    fun deleteDriver(@PathVariable("id") id: String): Responses {
        val driver = driverRepository.findDriverById(id)

        return if (driver != null) {
            driver.let { driverRepository.delete(it) }
            Responses("OK", listOf(null))
        } else {
            Responses("FAILED", listOf(null))
        }
    }
}